homeassistant/components/tradfri/coordinator.py

```diff
@@ -0,0 +1,89 @@
+"""Tradfri DataUpdateCoordinator."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import timedelta
+import logging
+from typing import Any
+
+from aiocoap.error import NetworkError
+import async_timeout
+from pytradfri.command import Command
+from pytradfri.device import Device
+from pytradfri.error import PytradfriError
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import SCAN_INTERVAL, TIMEOUT_API
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TradfriDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """Coordinator to manage data for a specific Tradfri device."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        api: Callable[[Command | list[Command]], Any],
+        device: Device,
+    ) -> None:
+        """Initialize device coordinator."""
+        self.api = api
+        self.device = device
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"Update coordinator for {device}",
+            update_interval=timedelta(seconds=SCAN_INTERVAL),
+            update_method=self._update_data,
+        )
+
+    @callback
+    def _observe_update(self, device: Device) -> None:
+        """Update the coordinator for a device when a change is detected."""
+        self.async_set_updated_data(data=device)
+
+    async def _async_run_observe(self, cmd: Command, device: Device = None) -> None:
+        """Run observe in a coroutine."""
+        try:
+            await self.api(cmd)
+        except PytradfriError as err:
+            _LOGGER.warning("Observation failed, trying again", exc_info=err)
+            self._async_start_observe(device=device)
```

Comment:
Probably refresh the coordinator here instead.

Maybe we can store the exception as a coordinator attribute and check this attribute first in the coordinator update method. If the attribute is set we raise it as an exception.

Maybe also set a lower update interval on error so we get a faster retry?

Reply:
I've refactored the code to try to accomodate to how I've interpreted these ideas.